evm/v0.5/contracts/dev/PrepaidAggregator.sol

```diff
@@ -0,0 +1,242 @@
+pragma solidity 0.5.0;
+
+import "../Median.sol";
+import "../vendor/Ownable.sol";
+import "../vendor/SafeMath.sol";
+import "./SafeMath128.sol";
+import "../interfaces/LinkTokenInterface.sol";
+
+/**
+ * @title The PrepaidAggregator handles aggregating data pushed in from off-chain.
+ */
+contract PrepaidAggregator is Ownable {
+  using SafeMath for uint256;
+  using SafeMath128 for uint128;
+
+  struct OracleStatus {
+    bool enabled;
+    uint128 withdrawable;
+    uint256 lastReportedRound;
+  }
+
+  struct Round {
+    uint64 maxAnswers;
+    uint64 minAnswers;
+    uint128 paymentAmount;
+    int256[] answers;
+  }
+
+  int256 public currentAnswer;
+  uint256 public currentRound;
+  uint128 public paymentAmount;
+  uint64 public oracleCount;
+  uint64 public maxAnswerCount;
+  uint64 public minAnswerCount;
+  uint256 public updatedHeight;
+  uint128 public availableFunds;
+  uint128 public allocatedFunds;
+
+  LinkTokenInterface private LINK;
+  mapping(address => OracleStatus) private oracles;
+  mapping(uint256 => Round) private rounds;
+
+  event NewRound(uint256 indexed number, address indexed startedBy);
+  event AnswerUpdated(int256 indexed current, uint256 indexed round);
+  event AvailableFundsUpdated(uint256 indexed amount);
+
+  constructor(address _link, uint128 _paymentAmount) public {
+    LINK = LinkTokenInterface(_link);
+    setPaymentAmount(_paymentAmount);
+  }
+
+  function updateAnswer(uint256 _round, int256 _answer)
+    public
+    ensureValidRoundId(_round)
+    validateOracleRound(_round)
+  {
+    startNewRound(_round);
+    recordAnswer(_answer, _round);
+    updateRoundAnswer(_round);
+    payOracle(_round);
+    deleteRound(_round);
+  }
+
+  function addOracle(address _oracle)
+    public
+    onlyOwner()
+    ensureNotEnabledAddress(_oracle)
+  {
+    require(oracleCount < 42, "cannot add more than 42 oracles");
+    oracles[_oracle].enabled = true;
+    oracleCount += 1;
+    setAnswerCountRange(minAnswerCount + 1, maxAnswerCount + 1);
+  }
+
+  function removeOracle(address _oracle)
+    public
+    onlyOwner()
+    ensureEnabledAddress(_oracle)
+  {
+    oracles[_oracle].enabled = false;
+    oracleCount -= 1;
+
+    uint64 min = minAnswerCount;
+    uint64 max = maxAnswerCount;
+    if (min > 0) {
+      min = min - 1;
+    }
+    if (max > 0) {
+      max = max - 1;
+    }
+    setAnswerCountRange(min, max);
+  }
+
+  function setPaymentAmount(uint128 _newAmount)
+    public
+    onlyOwner()
+  {
+    paymentAmount = _newAmount;
+  }
+
+  function setAnswerCountRange(uint64 _min, uint64 _max)
+    public
+    onlyOwner()
+    ensureValidRange(_min, _max)
+  {
+    minAnswerCount = _min;
+    maxAnswerCount = _max;
+  }
+
+  function updateAvailableFunds()
```

Comment:
Not necessary, but it would be nice if we had an `onTokenTransfer` method that simply called this so that we don't have to do 2 transactions when funding.

Reply:
Wasn't sure if it was worth the complexity, but I think you're right.

Reply:
This PR is pretty long running and `onTokenTransfer` is a nice to have, so I'd like to save that for another PR.